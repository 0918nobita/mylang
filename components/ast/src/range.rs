use std::fmt;

use serde::{Deserialize, Serialize};

use super::pos::Pos;

#[derive(Clone, Default, Serialize, Deserialize)]
pub struct Range {
    pub start: Pos,
    pub end: Pos,
}

impl Range {
    pub fn new(start: Pos, end: Pos) -> Self {
        Range { start, end }
    }
}

impl fmt::Debug for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

impl fmt::Display for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

impl From<Pos> for Range {
    fn from(pos: Pos) -> Self {
        Self {
            start: pos.clone(),
            end: pos,
        }
    }
}

pub trait Locatable {
    fn locate(&self) -> Range;
}
