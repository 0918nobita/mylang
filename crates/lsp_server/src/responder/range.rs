use mylang_token::Locatable;
use serde_json::Value as JsonValue;

pub fn locatable_to_lsp_range<L>(locatable: &L) -> JsonValue
where
    L: Locatable,
{
    let range = locatable.locate();
    serde_json::to_value(range).expect("Failed to convert locatable to json")
}
