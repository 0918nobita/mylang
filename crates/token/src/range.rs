use std::fmt;

use serde::{Deserialize, Serialize};

use super::pos::Pos;

/// ソースコード中の範囲
#[derive(Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct Range {
    /// 始端
    start: Pos,
    /// 終端
    end: Pos,
}

impl Range {
    /// 新しい範囲を生成して返す
    pub fn new(start: Pos, end: Pos) -> Self {
        Range { start, end }
    }

    /// 終端の参照を返す
    pub fn end_ref(&self) -> &Pos {
        &self.end
    }

    /// 所有権をムーブして終端を返す
    pub fn end(self) -> Pos {
        self.end
    }

    /// 終端を別の範囲の終端に設定する
    ///
    /// ```
    /// use token::range;
    ///
    /// let mut former = range!(0;0, 0;5);
    ///
    /// let latter = range!(0;7, 0;10);
    /// former.concat(latter);
    ///
    /// let expected = range!(0;0, 0;10);
    /// assert_eq!(former, expected);
    /// ```
    pub fn concat(&mut self, other: Range) {
        self.end = other.end;
    }
}

impl fmt::Debug for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

impl fmt::Display for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}-{}", self.start, self.end)
    }
}

impl From<Pos> for Range {
    /// 始端と終端が同じ位置の範囲に変換する
    fn from(pos: Pos) -> Self {
        Self {
            start: pos.clone(),
            end: pos,
        }
    }
}
