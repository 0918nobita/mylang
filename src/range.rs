use std::fmt;

use super::pos::Pos;

#[derive(Clone, Default)]
pub struct Range {
    start: Pos,
    end: Pos,
}

impl fmt::Debug for Range {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}-{:?}", self.start, self.end)
    }
}

pub trait Locatable {
    fn get_range(&self) -> Option<Range>;
}
