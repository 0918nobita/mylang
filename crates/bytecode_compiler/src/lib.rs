//! 抽象構文木→バイトコード変換器

use mylang_ast::{Expr, Stmt};
use mylang_bytecode::Inst;

fn expr_to_bytecode(expr: &Expr) -> Vec<Inst> {
    match expr {
        Expr::I32Lit(_, i) => vec![Inst::I32Const(*i)],
        Expr::Add(lhs, rhs) => {
            let mut insts = expr_to_bytecode(lhs);
            insts.extend(expr_to_bytecode(rhs));
            insts.push(Inst::I32Add);
            insts
        }
        Expr::StrLit(_, s) => vec![Inst::StrConst(s.clone())],
    }
}

fn stmt_to_bytecode(stmt: &Stmt) -> Vec<Inst> {
    match stmt {
        Stmt::PrintI32(_, ref expr) => {
            let mut insts = expr_to_bytecode(expr);
            insts.push(Inst::PrintI32);
            insts
        }
        Stmt::PrintStr(_, ref expr) => {
            let mut insts = expr_to_bytecode(expr);
            insts.push(Inst::PrintStr);
            insts
        }
    }
}

pub fn ast_to_bytecode(ast: &[Stmt]) -> Vec<Inst> {
    let mut insts = vec![];
    for stmt in ast {
        insts.extend(stmt_to_bytecode(stmt));
    }
    insts
}

#[cfg(test)]
mod tests {
    use mylang_ast::{Expr, Stmt};
    use mylang_token::range;

    use crate::{expr_to_bytecode, stmt_to_bytecode};

    #[test]
    /// 整数リテラルから生成されるバイトコードが変化していないことを確認する
    fn i32_lit_to_bytecode() {
        insta::assert_debug_snapshot!(expr_to_bytecode(&Expr::I32Lit(range!(0;0), 3)));
    }

    /// 文字列リテラルから生成されるバイトコードが変化していないことを確認する
    #[test]
    fn str_lit_to_bytecode() {
        let bytecode = expr_to_bytecode(&Expr::StrLit(range!(0;0,0;4), "Hello".to_owned()));
        insta::assert_debug_snapshot!(bytecode);
    }

    #[test]
    /// 整数リテラル同士の足し算から生成されるバイトコードが変化していないことを確認する
    fn addition_to_bytecode() {
        let addition = Expr::Add(
            Box::new(Expr::I32Lit(range!(0;0), 3)),
            Box::new(Expr::I32Lit(range!(0;4,0;4), 4)),
        );
        let bytecode = expr_to_bytecode(&addition);
        insta::assert_debug_snapshot!(bytecode);
    }

    #[test]
    /// print_i32 文から生成されるバイトコードが変化していないことを確認する
    fn print_i32_to_bytecode() {
        let print_i32 = Stmt::PrintI32(range!(0;0,0;8), Expr::I32Lit(range!(0;10), 3));
        let bytecode = stmt_to_bytecode(&print_i32);
        insta::assert_debug_snapshot!(bytecode);
    }

    #[test]
    /// print_str 文から生成されるバイトコードが変化していないことを確認する
    fn print_str_to_bytecode() {
        let print_str = Stmt::PrintStr(
            range!(0;0,0;8),
            Expr::StrLit(range!(0;10,0;23), "Hello, world!".to_owned()),
        );
        let bytecode = stmt_to_bytecode(&print_str);
        insta::assert_debug_snapshot!(bytecode);
    }
}
