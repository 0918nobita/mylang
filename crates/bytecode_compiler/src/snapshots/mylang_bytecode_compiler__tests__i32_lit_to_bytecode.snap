---
source: crates/bytecode_compiler/src/lib.rs
assertion_line: 51
expression: "expr_to_bytecode(&Expr::I32Lit(range!(0 ; 0), 3))"

---
[
    I32Const(
        3,
    ),
]
